tools/Solidity.g4

```diff
@@ -0,0 +1,482 @@
+  : (pragmaDirective | importDirective | structDefinition | enumDefinition | contractDefinition)* EOF ;
+
+pragmaDirective
+  : 'pragma' pragmaName pragmaValue ';' ;
+
+pragmaName
+  : identifier ;
+
+pragmaValue
+  : version | expression ;
+
+version
+  : versionConstraint versionConstraint? ;
+
+versionConstraint
+  : versionOperator? VersionLiteral ;
+
+versionOperator
+  : '^' | '~' | '>=' | '>' | '<' | '<=' | '=' ;
+
+importDirective
+  : 'import' StringLiteralFragment ('as' identifier)? ';'
+  | 'import' ('*' | identifier) ('as' identifier)? 'from' StringLiteralFragment ';'
+  | 'import' '{' importDeclaration ( ',' importDeclaration )* '}' 'from' StringLiteralFragment ';' ;
+
+importDeclaration
+  : identifier ('as' identifier)? ;
+
+contractDefinition
+  : 'abstract'? ( 'contract' | 'interface' | 'library' ) identifier
+    ( 'is' inheritanceSpecifier (',' inheritanceSpecifier )* )?
+    '{' contractPart* '}' ;
+
+inheritanceSpecifier
+  : userDefinedTypeName ( '(' expressionList? ')' )? ;
+
+contractPart
+  : stateVariableDeclaration
+  | usingForDeclaration
+  | structDefinition
+  | modifierDefinition
+  | functionDefinition
+  | eventDefinition
+  | enumDefinition ;
+
+stateVariableDeclaration
+  : typeName
+    ( PublicKeyword | InternalKeyword | PrivateKeyword | ConstantKeyword | ImmutableKeyword | overrideSpecifier )*
+    identifier ('=' expression)? ';' ;
+
+overrideSpecifier : 'override' ( '(' userDefinedTypeName (',' userDefinedTypeName)* ')' )? ;
+
+usingForDeclaration
+  : 'using' identifier 'for' ('*' | typeName) ';' ;
+
+structDefinition
+  : 'struct' identifier
+    '{' ( variableDeclaration ';' (variableDeclaration ';')* )? '}' ;
+
+modifierDefinition
+  : 'modifier' identifier parameterList? ( VirtualKeyword | overrideSpecifier )* block ;
+
+functionDefinition
+  : functionDescriptor parameterList modifierList returnParameters? ( ';' | block ) ;
+
+functionDescriptor
+  : 'function' ( identifier | ReceiveKeyword | FallbackKeyword )?
+  | ConstructorKeyword
+  | FallbackKeyword
+  | ReceiveKeyword ;
+
+returnParameters
+  : 'returns' parameterList ;
+
+modifierList
+  : ( modifierInvocation | stateMutability | ExternalKeyword
+    | PublicKeyword | InternalKeyword | PrivateKeyword | VirtualKeyword | overrideSpecifier )* ;
+
+modifierInvocation
+  : identifier ( '(' expressionList? ')' )? ;
+
+eventDefinition
+  : 'event' identifier eventParameterList AnonymousKeyword? ';' ;
+
+enumDefinition
+  : 'enum' identifier '{' enumValue? (',' enumValue)* '}' ;
+
+enumValue
+  : identifier ;
+
+parameterList
+  : '(' ( parameter (',' parameter)* )? ')' ;
+
+parameter
+  : typeName storageLocation? identifier? ;
+
+eventParameterList
+  : '(' ( eventParameter (',' eventParameter)* )? ')' ;
+
+eventParameter
+  : typeName IndexedKeyword? identifier? ;
+
+variableDeclaration
+  : typeName storageLocation? identifier ;
+
+typeName
+  : elementaryTypeName
+  | userDefinedTypeName
+  | mapping
+  | typeName '[' expression? ']'
```

Comment:
Would it not be better to still have `arrayTypeName`? I could imagine that the grammar is then easier to use, but I'm not sure.

Reply:
Yes, I tried it. It complains about left-recursiveness.